catkin_ws/src/calqui/scripts/goto_server.py

```diff
@@ -0,0 +1,138 @@
+        self._fb_pub = rospy.Publisher("/goto_place/feedback", GoToPlaceFeedback, queue_size=10)
+        self._odom_sub = rospy.Subscriber("/odom", Odometry, self.on_odom)
+        self._current_pose = Pose2D()
+        self._goal_pose = Pose2D()
+        self._as.start()
+        
+        rospy.loginfo('Simple Action Server has been started')
+
+    def on_shutdown(self):
+        rospy.loginfo('stopping vel')  
+        self.set_vel(0,0)
+
+    def on_odom(self, msg):
+        global roll, pitch, yaw
+        orientation_q = msg.pose.pose.orientation
+        orientation_list = [orientation_q.x, orientation_q.y, orientation_q.z, orientation_q.w]
+        (roll, pitch, yaw) = euler_from_quaternion (orientation_list)
+        # rospy.loginfo('yaw:{}, x:{} y:{} z:{}'.format(round(yaw,2), round(msg.pose.pose.position.x,2),round(msg.pose.pose.position.y,2),round(msg.pose.pose.position.z,2) ))
+        self._current_pose.x = msg.pose.pose.position.x
+        self._current_pose.y = msg.pose.pose.position.y
+        self._current_pose.theta = yaw
+        
+        # rospy.loginfo('current pose is {}'.format(self._current_pose))
+        fb = GoToPlaceFeedback()
+        fb.distance_to_goal = self.distance_to_goal()
+        self._fb_pub.publish(fb)
+        
+
+    def set_vel(self, linear_x, angular_z):
+        rospy.loginfo('publishing vel')
+        msg = Twist()
+        msg.linear.x = linear_x
+        msg.linear.y = 0
+        msg.linear.z = 0
+        msg.angular.x = 0
+        msg.angular.y = 0
+        msg.angular.z = angular_z
+        self._vel_pub.publish(msg)
+
+    def distance_to_goal(self):
+        return math.sqrt(math.pow((self._goal_pose.y - self._current_pose.y),2) + math.pow((self._goal_pose.x - self._current_pose.x),2))
+
+    # absolute angle to the goal
+    def angle_to_goal(self):
+        # avoid division by 0
+        if self._current_pose.x == 0:
+            return 0 
+        else:
+            return numpy.arctan2((self._goal_pose.y - self._current_pose.y), (self._goal_pose.x - self._current_pose.x))
+
+    # how far are we of pointing to the goal?
+    def deviation(self):
+        return self._current_pose.theta - self.angle_to_goal()
+
+    # do we consider that we reached the goal
+    def goal_was_reached(self):
+        return self._current_pose and self._goal_pose and self.distance_to_goal() < 0.5
+
+    def calculate_angular_speed(self):
+        if numpy.abs(self.deviation()) < 0.1 or self.goal_was_reached():
+            rotation_speed = 0
+        else:
+            rotation_speed = -0.1 * self.deviation()
+
+        rospy.loginfo('rotation_speed {}'.format(rotation_speed))
+        rospy.loginfo('angle_to_goal {}, current_angle {}'.format(self.angle_to_goal(), self._current_pose.theta))
+        return rotation_speed
+
+    def calculate_linear_speed(self, angular_speed):
+        if self.goal_was_reached() or numpy.abs(self.deviation()) > 0.5:
+            return 0
+        return 0.5
```

Comment:
you can also try a proportional controller here (i.e. K*distance_to_goal).

Intuitively, what happens is that the robot moves faster when it's further away, and starts moving slower when it's reaching the goal ("avoiding" then to overshoot).

A real robot has of course a maximum allowed speed, so it would be good to check in turtlebot docs that number (probably the turtlebot controller is already checking that and setting max speed if the requested velocity is bigger than allowed, so that check might not be needed).